packages/SwingSet/docs/garbage-collection.md

```diff
@@ -358,6 +358,14 @@ At this point, the kref is only referenced in the queued `retireImport` action a
 
 If the last importing vat had previously called `syscall.retireImport`, there will be no subscribers, but the kernel object data will still be present (a general invariant is that the exporting clist entry and the kernel object table entry are either both present or both missing). A `dispatch.retireExport` will be on the queue for the exporting vat, but it has not yet arrived (otherwise it would be illegal for the exporting vat to call `syscall.retireExport`). That `dispatch.retireExport` GC action will be nullified during `processOneGCAction` because its work was already performed by the exporter's `syscall.retireExport`.
 
+## syscall.abandonExport processing
+
+During vat upgrade, the last delivery made to the old version is a special `dispatch.stopVat()`. This instructs the soon-to-be-deleted worker to destroy anything that will not survive the upgrade. All Remotables will be abandoned here, because they live solely in RAM, and the RAM heap does not survive. All non-durable virtual objects will also be abandoned, since only durable ones survive. We may also drop otherwise-durable objects which were only referenced by abandoned non-durable objects.
+
+During this phase, the vat performs a `syscall.abandonExports()` with all the abandoned vrefs. The kernel reacts to this in roughly the same way it reacts to the entire vat being terminated (which it is, sort of, at least the heap is being terminated). Each vref is deleted from the exporting vat's c-list, because *that* vat isn't going to be referencing it any more. The kernel object table is updated to clear the `owner` field: while the kernel object retains its identity, it is now orphaned, and any messages sent to it will be rejected (by the kernel) with a "vat terminated" error.
```

Comment:
Re: "a `syscall.abandonExports()` with all the abandoned vrefs": presumably Liveslots could issue multiple such calls?  My read of the code suggests there's no reason from the kernel's perspective it has to be all at once.  I can imagine batching if the number of severed exports is very large, or possibly different cleanup passes that discover different things (e.g., stuff in non-durable stores, then stuff in non-durable VOs, then stuff in unreferenced durable VOs, etc.).

Also, is `syscall.abandonExports` limited strictly to use during vat stoppage?  It seems like it is closely related to the `vatPowers.disavow` operation (indeed, it's not clear to me how they could be any different).

Reply:
It could make multiple calls, yeah. It's more efficient to do larger batches, and I'll probably make `stopVat` do a single large call, because I'm more likely to get it right that way (I'm already having problems with `stopVat` not being aware of what is dropped by the `bringOutYourDead` call that it makes, and then `abandonExports` does something bad).

It *could* be called at other times, but I think `stopVat` is the only appropriate time because we're basically yanking the rug out from userspace, we're breaking their promise to make that previously-exported object available, and if we ever allow userspace to run after doing `stopVat`, they're likely to become confused or cause significant confusion.

`disavow` is for imports; it says "I have this Presence here, and I know that means I'm holding a reference, but pretend that I just dropped it and drop/retire the import". It's a different sort of rug-pull than the exports that `syscall.abandonExport` does. Abandoning an export is a tiny bit more serious than disavowing an import: the import is depended upon by userspace code (if it tries to do something with the Presence, things will break), whereas the export is depended upon by the kernel and other vats, so `abandonExport` lets you surprise more than just yourself. That said, neither is really meant for userspace control, and `vatPowers.disavow` is mainly just for unit tests.

Reply:
Ah, I hadn't twigged to "disavow import" vs. "disavow export".  Naturally the latter is more radical, but vat code always has the option to act as if an export has been disavowed by simply throwing unconditionally in response to everything send to it (which means exposing `abandonExports` doesn't really do anything useful, so, yeah, let's not).